frontend/src/utils/date.ts

```diff
@@ -0,0 +1,35 @@
+import dayjs from "dayjs"
+import timezone from "dayjs/plugin/timezone"
+import utc from "dayjs/plugin/utc"
+import { isoDateTimePattern } from "./validation"
+
+/**
+ * Format given date to specified format based relative to user's timezone.
+ * @param date - The date string to format (expected to be in ISO 8601 date format).
+ * @returns Formatted date string or "Invalid Date" if the input is not valid.
+ *
+ * Examples:
+ *  Input: "2024-09-23T12:00:00Z"
+ *  Expected Output: "Sep 23, 2024 08:00AM" (if user's in America/New_York timezone)
+ */
+
+// dayjs plugins
+/* eslint-disable */
+dayjs.extend(utc)
+dayjs.extend(timezone)
+/* eslint-enable */
+
+export const formatDate = (date: string): string => {
+  const errorMessage = "Invalid Date"
+
+  if (typeof date !== "string" || !isoDateTimePattern.value.test(date))
+    return errorMessage
```

Comment:
question: in this situation do we perhaps want to throw an exception? The reason I ask is this seems like a truly exceptional case -- ie. bad data passed in. In this situation, you don't want the caller of this utility to accidentally ignore a string message. You want them to crash, find the bug, and fix it. Also known as fail fast. So either they have to wrap this call in a try/catch, or they have to make sure they pass the right data in. Otherwise you risk GIGO.

Reply:
Thank you for pointing this out. Yes, we'd want the caller to not worry about manually catching error when calling this function. I'll revise as you suggested and throw exception on these error cases.

Reply:
Refactored to throw three types of errors:

```typescript

export const INVALID_DATE_TYPE_ERROR_MESSAGE = "Date must be a string"
export const ISO_DATE_FORMAT_ERROR_MESSAGE = "Date must be in ISO 8601 format"
export const DATE_PARSING_ERROR_MESSAGE = "Error parsing the date"

export const formatISODateToUserTimezone = (date: string): string => {
  if (typeof date !== "string") {
    throw new Error(INVALID_DATE_TYPE_ERROR_MESSAGE)
  }

  if (!isoDateTimePattern.value.test(date)) {
    throw new Error(ISO_DATE_FORMAT_ERROR_MESSAGE)
  }

  // Unsafe access to 'error' typed value is handled by dayjs' isValid function
  /* eslint-disable */
  const parsedDate = dayjs.utc(date)
  if (!parsedDate.isValid()) {
    throw new Error(DATE_PARSING_ERROR_MESSAGE)
  }

  const userTimezone = dayjs.tz.guess()
  return parsedDate.tz(userTimezone).format("MMM DD, YYYY hh:mm A")
  /* eslint-enable */
}
```

Reply:
If the caller were passing anything besides a string wouldn't the code fail to compile bc of a TypeError?

Reply:
I don't love the throwing of errors  in a case like this.  This is a tiny utility function and a thrown error could crash the entire application.  It would be very easy for a programmer to pass in a date string like 2024-11-26 without realizing how specialized this funciton is, and we wouldn't want the entire app to crash.  Would we?  I don't know.  I am a bit nervous about this now -- whether this could lead to subtle bugs down the road.  Again, I would be happier if this funciton were just "convertDatabaseDataTimeToStandardFormat" so we know the exact use case where this is a useful function.

Reply:
Yeah throwing is probably not the right thing here. An error string makes more sense I think.

Reply:
If this function is called on just single data, then yes the code will fail to compile.
But, the way I see how this function will be primarily used for is formatting multiple data (fetched from the db) with dateTime field and rendering them.

Suppose we have following data:
```
[
  {
    id: "92bc71bf-f42c-4b56-b55d-fddaf4633550",
    text: "How is your work-life balance?",
    created_at: "2024-09-22T18:20:53.734830",
  },
  {
    id: "736381f5-becf-4783-8f04-98250e69c0b3",
    text: "Is communication effective?",
    created_at: 12345, // notice this is not a string
  },
  {
    id: "df1e0d27-4f77-4d6b-b997-2b07d892db17",
    text: "How satisfied are you with the work environment?",
    created_at: "2024-09-22T12:30:53.734830",
  }
]
```

And we're rendering some component using this data:
```
data.map(d => (
  <>
    <>{d.id}</>
    <>{d.text}</>
    <>{d.created_at}</>
  </>
))
```

The code will still compile and render the component successfully.

It is unlikely that the dateTime field will be in non-string format, since the data will be converted into JSON when fetched. But I wanted to address the possibility of data containing non-string dataTime value.

Reply:
So folks, I think the principle is: if it's a coding error, fail fast. i.e. crash the entire app -- so that the dev writing the code (or the E2E tests) find the error. The problem with returning a string message is now you're relying on either the language to catch a mistype (and a TypeError is much less obvious than a custom one), or for the code to gracefully handle the unexpected return. (We aren't explicitly scanning for the error message, right?)

The basic idea is it's better to just crash and find/fix the bug fast, than to have a subtle hidden bug lurk and then spend days trying to find it.

Reply:
Now, if it's a user error (e.g. entered wrong data, etc) then of course, handle gracefully and do not crash.